# Analysis of RNA Sequencing Data Using CLC Genomics Workbench and R: An Example of Identifying genetic expressions and variations in comparative experiments

## introduction

## preparation

### dataset

#### objective and background

#### gene types and treatments

#### organization of files and nomination

### CLC workbench

#### requirements

#### download

#### plugin installation

#### Licenses

#### CLCBio Genomics Workbench Server

### IPA

#### requirements

#### download

#### Licenses

### R

#### requirements
- Operating System: R can be installed on Windows, macOS, and Linux platforms.
- Hardware Requirements: No specific hardware requirements. Performance improves with better hardware.
- Memory Requirements: 2 GB of RAM minimum recommended, 4 GB of RAM recommended for handling larger datasets.

#### download
To use R for data analysis, you will need to install both R and RStudio, a popular Integrated Development Environment (IDE) that makes using R easier. download, installation instruction can be found:https://posit.co/download/rstudio-desktop/

#### package installation
R packages are collections of functions and data sets developed by the community that enhance R’s functionality, allowing you to perform additional types of analysis or improve ease of use. For example, ```ggplot2``` enhances graphical capabilities, ```dplyr``` offers advanced data manipulation, and tidyr helps in organizing data. 

![RStudio Interface: Install and load package in console](images_protocol/1718313263255.jpg)

Installing packages in R is straightforward:

- Open RStudio: After installation, open RStudio. The interface is divided into several panels:

  - Console: Where you type commands and see output.
  - Environment/History: Shows current variables in the workspace and command history.
  - Files/Plots/Packages/Help: Access to files, visual outputs, package management, and help files.
 
- Installing Packages: To install an R package, use the ```install.packages()``` function in the Console. For example:

 ```
 install.packages("ggplot2") # Installs the ggplot2 package for data visualization
 ```
- Loading Packages: After installation, load a package into your session to use its functions:

 ```
 library(ggplot2) # Loads the ggplot2 package for use
 ```

#### Licenses

R and RStudio are free software. R is released under the GNU General Public License, and RStudio comes under the Affero General Public License. These licenses allow you to use, modify, and distribute the software freely.

## Method

### RNA Sequencing Data Analysis Workflow

### Data preprocessing

#### Import FASTQ Reads to CLC Genomics Workbench

#### Quality Control of Sequencing Reads

#### Trim Reads

#### Import Metadata Table

#### Reads Mapping

#### Automate the process using Workflow

#### Export Data

### General downstream analysis

#### with CLC workbench

#### Create a PCA Plot

#### Differential Expression

#### Create a Heatmap

#### Create a Venn Diagram

### IPA analysis

### Customized downstream analysis with R

#### get sample codes from GitHub

- For users familiar with GitHub:

   Clone the Repository: To get the sample code, clone the repository from GitHub using the following command in your terminal or command prompt:
   
   ```
   git clone https://github.com/maxianren/bioinformatics_analysis.git
   ```
   
   This will create a local copy of the repository on your machine, where you can leverage the code to create your own scripts and potentially contribute updates or improvements back to the repository.

- For users who do not use GitHub:

  - Download the Code:
   ![GitHub: Download the code](images_protocol/1718316248835.jpg)
   
  - Go to https://github.com/maxianren/bioinformatics_analysis.
  - Click on the Code button and select Download ZIP.
  - Save the ZIP file to your computer and extract it to a desired location.
 

#### explanation and rational of the code
Once you have the codebase on your local machine, here’s what you will find inside:

- Objective of the Code:
 The provided scripts are designed to perform customized data analysis on results generated from CLC Genomics Workbench. The main objectives are:
  - Data Transformation: Convert raw data into a format suitable for analysis and visualization.
  - Plot Generation: Create informative visualizations to compare treatments within the same gene type. Example plots include heatmaps for visual comparison, box plots for statistical distribution, and GO analysis for understanding gene functions and pathways.
 ![Plot Generation Process](images_protocol/1718322571364.jpg)
 
- Folder Structure:
  - data/: This directory should contain the input data, which is generated by the CLC Workbench.
  - out/: This is where the output of the program is stored, including any plots or processed data files.
  - scripts/: Contains the R scripts used for analysis. There are four main files:
    - install.R: This script is used for installing necessary R packages.
    - data_format.R: Prepares and formats the data for analysis.
    - draw.R: Contains functions to generate plots.
    - main_ops.R: Main operations script that calls functions from other scripts to process data and generate plots.
  - bioinformatics_analysis.Rproj: An R project file that can be opened with RStudio to manage the entire project easily.


#### allocate input files and output path
To organize your data efficiently, you may create specific folders within the data/ and out/ directories:

- Input Files:
  - Create a project-specific folder within the data/ directory. For example, for an LPS-related project, create an LPS/ folder.
  - Inside the LPS/ folder, create a genes/ subfolder to store gene data for each experimental sample.
  - Place all input files in the appropriate directories. Ensure that these files are correctly formatted and named for easy reference by the scripts.
 ![Input Folder](images_protocol/1718329782355.jpg)
- Output Files:
  - Similarly, create a corresponding project folder within the out/ directory to store all outputs generated by your analyses.
  - This organized structure helps in tracking the analysis outputs related to specific input sets and projects.
 ![Output Folder](images_protocol/1718329786804.jpg)
  
#### run the code to generate desired plots
To execute the analysis and generate plots:

- Open the Project:
  - Launch RStudio.
  - Open bioinformatics_analysis.Rproj to initiate the R project environment.
  - Manually set working directory if the working directory is not set properly:Session > Set Working Directory > To Project Directory.
- Prepare the Scripts:
  - Open main_ops.R. This script orchestrates the workflow, calling functions from other scripts.
  - Ensure main_ops.R includes references to other necessary scripts (install.R, data_format.R, draw.R) using source() function calls.
  - In main_ops.R, specify paths for input data and output files.
  ![Include reference scripts and define input and ouput locations](images_protocol/1718921091701.jpg)
  - Execute the relevant sections of the script by selecting the lines and pressing "Run" or Ctrl+Enter (or Cmd+Enter on Mac).
  - Confirm that paths and data are correctly recognized and reference scripts are correclt included in the console output.
  ![Include reference scripts and define input and ouput locations](images_protocol/1718921086289.jpg)
- Generate Plots:
  - Currently, the script includes commands to create box plots, perform GO analysis, and generate heatmaps.
  - Example to perform GO analysis:
    - Copy the following code:
      ```
      diff_expr_file = "diff_expr_WT - 5_g vs. PBS.csv" # Change as you need
      log_2_FC = 1 # Change as you need
      top_n = 10 # Change as you need
      
      control_group = str_remove_all(diff_expr_file, "diff_expr_|\\.csv")
      
      # plot
      go_plot <- GoAnalysisPlot$new(data_dir = data_dir,
                                    gene_folder = gene_folder,
                                    output = file.path(out_dir, sprintf("go_enrichment_%s.png", control_group)), 
                                    width = 2000, 
                                    height = 4000, 
                                    res = 300, 
                                    bg = "white")
      go_plot$draw(diff_expr_file = diff_expr_file, 
                   top_n = top_n, 
                   title = sprintf("Go Analysis: %s", control_group))
      ```
    - Adjust the parameters as needed, select the script lines, and run them to generate and save the plot in the specified out/ directory.
      - log_2_FC: It represents the log base 2 fold change threshold. It’s a criterion used to filter genes based on the magnitude of their expression changes. 
      - top_n: Specifies the number of top differentially expressed genes to consider in the analysis.
    ![Adjust parameters to generate plot](images_protocol/1718921839056.jpg)
    - The results should be like:
    ![Result](images_protocol/1718921839057.png)
  - Example to generate box plot:
    - Copy the following code:
      ```
      ## KO - 5_g vs. PBS ===============================
      # Var
      diff_expr_file = "diff_expr_WT - 5_g vs. PBS.csv" # Change as you need
      pattern = "(.*WT.*5_g.*|.*WT.*PBS.*)" # Change as you need
      top_n = 6 # Change as you need
      
      feature = "TPM"
      control_group = str_remove_all(diff_expr_file, "diff_expr_|\\.csv")
      
      # plot
      box_plot <- BoxPlot$new(data_dir = data_dir,
                              gene_folder = gene_folder,
                              output = file.path(out_dir, sprintf("box_plot_KO - 5_g vs. PBS.png", control_group)),
                              width = 3000, 
                              height = 2400, 
                              res = 300, 
                              bg = "white")
      box_plot$draw(diff_expr_file = diff_expr_file, 
                    title = sprintf("Box Plot %s", control_group),
                    pattern = pattern,
                    feature = feature)
      ```
    - Adjust the parameters as needed, select the script lines, and run them to generate and save the plot in the specified out/ directory.
      - diff_expr_file: specifies the filename containing the differential expression data. 
      - pattern: This is a regular expression pattern used to filter filenames or data within files. The pattern "(.*WT.*5_g.*|.*WT.*PBS.*)" matches filenames that include “WT” with “5_g” or “PBS,” specifying which subsets of the data to focus on.
      - top_n: Specifies the number of top differentially expressed genes to consider in the analysis.
    ![Adjust parameters to generate plot](images_protocol/1718922199380.jpg)
    - The results should be like:
    ![Result](images_protocol/1718922199381.png)
  - Example to perform heatmap:
    - Copy the following code:
      ```
      ## compare treatment KO - 100_g vs. PBS====
      # var
      diff_expr_file = "diff_expr_KO - 100_g vs. PBS.csv" # Change as you need
      pattern = "(.*KO.*100_g.*|.*KO.*PBS.*)" # Change as you need
      
      feature = "TPM"
      control_group = str_remove_all(diff_expr_file, "diff_expr_|\\.csv")
      
      # top 2Ns
      for (top_n in c(30, 100, 250)) {
        if (top_n < 50) {flag_row_name = T} else {flag_row_name = F}
       
        # plot
        heatmap_plot <- HeatmapPlot$new(data_dir = data_dir,
                                        gene_folder = gene_folder,
                                        output = file.path(out_dir, sprintf("heatmap_%s - top %d.png", control_group, top_n*2)), 
                                        width = 6000, 
                                        height = 4800, 
                                        res = 300, 
                                        bg = "white")
        heatmap_plot$draw(diff_expr_file = diff_expr_file,
                          column_title = sprintf("Gene Expression - %s, Sorted by Log2 fold change Top %d", control_group, top_n*2),
                          flag_row_name = flag_row_name,
                          pattern = pattern,
                          feature = feature)
      }
      ```
    - Adjust the parameters as needed, select the script lines, and run them to generate and save the plot in the specified out/ directory.
      - diff_expr_file: specifies the filename containing the differential expression data. 
      - pattern: This is a regular expression pattern used to filter filenames or data within files. The pattern "(.*WT.*5_g.*|.*WT.*PBS.*)" matches filenames that include “WT” with “5_g” or “PBS,” specifying which subsets of the data to focus on.
      - top_n: Specifies the number of top differentially expressed genes to consider in the analysis.
      - for (top_n in c(30, 100, 250)): This loop iterates over different values of top_n, allowing multiple analyses or visualizations for 30, 100, and 250 top differentially expressed genes, respectively.
    ![Adjust parameters to generate plot](images_protocol/1718922803411.jpg)
    - Unlike the previous code, this example demonstrates how to generate a heatmap that displays varying numbers of genes using a loop. The results should appear as follows:
    ![Result](images_protocol/1718922803412.png)
    ![Result](images_protocol/1718922803413.png)
    ![Result](images_protocol/1718922803414.png)

#### modify the code for customized plots
- Enhancing Plot Customization

  While previous examples demonstrate generating plots with minimal modifications, creating more sophisticated and tailored visualizations requires modifications to the draw.R script. Specifically, you can adjust the parameters within the draw method of each plotting class to suit your specific requirements.

- Creating New Type of Plots Leveraging Existing Classes and Methods

  To create different types of plots, you can utilize existing classes and methods that handle data transformation and set general plotting parameters. This approach ensures consistency in data handling and plot styling across different types of visualizations. To aid in further development and customization, consider reviewing the following diagram that outlines the architecture of the plotting classes and methods. This visualization will help you understand how different components interact and where to make effective enhancements:

  ![Architecture Diagram](images_protocol/1718938276336.jpg)

  The diagram should ideally provide a clear mapping of class inheritances, method dependencies, and how data flows through these components. Ensure that it highlights key areas where developers can inject custom code or modify existing functionalities to meet their visualization needs.
  
# Reference
